typing/ctype.ml

```diff
@@ -533,6 +533,25 @@ exception Non_closed of type_expr * bool
 let free_variables = ref []
 let really_closed = ref None
 
+(* [free_vars_rec] collects the variables of the input type
+   expression into the [free_variables] reference. It is used for
+   several different things in the type-checker, with the following
+   bells and whistles:
+   - If [really_closed] is Some typing environment, types in the environment
+     are expanded to check whether the apparently-free variable would vanish
+     during expansion.
+   - We collect both type variables and row variables, paired with a boolean
+     that is [true] if we have a row variable.
+   - We do not count "virtual" free variables -- free variables stored in
+     the abbreviation of an object type that has been expanded (we store
+     the abbreviations for use when displaying the type).
+
+  [free_vars] below returns a (variable * bool) list, and only passes a typing
+  environment if the ?env parameter is given.
```

Comment:
Just looking at the comment, I was not sure what `passes a typing environment` meant. Would it not be clearer to  just say `set really_closed`?

Reply:
I would like to find a compromise where I do not mention too much of this horrible global-mutable-variable-state-setting business, yet I document the main function where the complex stuff happens. I don't like your suggestion, but I will see if I can improve the text.

Reply:
I reformulated this part of the comment. It now says:

```ocaml
   The functions [free_vars] and [free_variables] below pass
   the typing environment as an optional [?env] parameter.
   [free_vars] returns a [(variable * bool) list], while
   [free_variables] drop the type/row information
   and only returns a [variable list].
```